src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1258,37 +1235,46 @@ internal async ValueTask<NpgsqlDataReader> ExecuteReader(CommandBehavior behavio
                         // to prevents a dependency on the thread pool (which would also trigger deadlocks).
                         // The WriteBuffer notifies this command when the first buffer flush occurs, so that the
                         // send functions can switch to the special async mode when needed.
-                        var sendTask = NonMultiplexingWriteWrapper(connector, async, finalCt);
+                        var sendTask = NonMultiplexingWriteWrapper(connector, async, writeCt);
 
-                        // The following is a hack. It raises an exception if one was thrown in the first phases
-                        // of the send (i.e. in parts of the send that executed synchronously). Exceptions may
-                        // still happen later and aren't properly handled. See #1323.
-                        if (sendTask.IsFaulted)
-                            sendTask.GetAwaiter().GetResult();
+                        try
+                        {
+                            // The following is a hack. It raises an exception if one was thrown in the first phases
+                            // of the send (i.e. in parts of the send that executed synchronously). Exceptions may
+                            // still happen later and aren't properly handled. See #1323.
+                            if (sendTask.IsFaulted)
+                                sendTask.GetAwaiter().GetResult();
+                        }
+                        catch (OperationCanceledException e)
```

Comment:
The breaking on write failures used to happen in NpgsqlWriteBuffer.Flush, what's the reason for this change? (commented there as well

Reply:
Explained in https://github.com/npgsql/npgsql/pull/3191#discussion_r499712591